proposals/scalable-robust-operator.md

```diff
@@ -0,0 +1,284 @@
+    of `ssh`. This script, before executing the command provided by `mpirun`,
+    transfers a file containing a mapping of pod names to IPs and appends it to
+    the worker’s `/etc/hosts`.
+    
+    Note: The v1 controller no longer copies the pod-to-IP mapping file. The
+    OpenMPI implementation does the [routing](https://www.open-mpi.org/faq/?category=tcp#tcp-routability-1.3).
+  - The `hostfile` for `mpirun`, listing the worker pod names and number of
+    slots (which could be the number of CPUs or GPUs). This list is built
+    programmatically.
+2. Creates a ServiceAccount+Role+RoleBinding for the launcher, which allow it
+  to:
+  - get/list/watch on Pods
+  - do pods/exec
+  This allows the launcher Pod to obtain details of the worker Pods and start
+  the process managers on them.
+3. If configured, it creates a Volcano PodGroup
+4. Creates a StatefulSet for workers (plain pods in v1). The Pod template
+  includes:
+  - mounts for the ConfigMap.
+  - `sleep` as command.
+5. Creates launcher Job (plain pod in v1). The Pod template includes:
+  - An init container, `kubectl-delivery`, described below.
+  - Environment variables for:
+    - replacing `ssh` for `kubexec.sh`
+    - the `hostfile` location
+  - Volumes for:
+    - The ConfigMap
+    - Sharing files from `kubectl-delivery`
+
+The launcher Job, as previously mentioned, contains an init container:
+`kubectl-delivery`. This is a Kubernetes controller that watches pods. It does
+the following:
+
+1. Copy kubectl from the image into the volume shared with the main container.
+2. Wait for all Pods in the hostfile to be running
+3. Generates a file mapping pod name to IP, in `/etc/hosts` format.
+
+To update the status of an MPIJob, the controller uses the status of the
+launcher Job. That is, when the launcher Job fails or succeeds, it’s status is
+copied to the MPIJob. In v1, it bases the status on the termination condition of
+the Pod.
+
+### Analysis
+
+The above architecture for MPI Jobs has the following problems:
+
+- It doesn’t scale as we increase the number of workers, belonging to the same
+  or different jobs.
+  - Due to the use of `kubectl exec`, every worker spawn goes through
+    kube-apiserver. `mpirun` starts a daemon in each worker
+    (like [`orted`](https://www.open-mpi.org/doc/v3.0/man1/orted.1.php)).
+    This process handles the worker-to-worker communication, which happens
+    without the intervention of kube-apiserver. However, the `exec` connection
+    stays up for control during the entirety of the job.
+  - The `kubectl-delivery` controller does a full cache sync to be able to watch
+    Pods. This startup penalty increases with the number of pods in the cluster
+    and has to be paid for every job. The API calls also causing additional
+    stress on the apiserver as the number of launchers increases.
+  - The launcher pod has execution permissions on any other Pod in the
+    namespace, which can be a security concern.
```

Comment:
The launcher has only the execution permission on the worker pods belong it's job, you can find it at [code](https://github.com/kubeflow/mpi-operator/blob/master/pkg/controllers/v1/mpi_job_controller.go#L1222)

Reply:
Ah, missed that. However, that doesn't scale as the number of workers increase (k8s objects have a size limit).

Reply:
> Ah, missed that. However, that doesn't scale as the number of workers increase (k8s objects have a size limit).

It does. We introduced this feature in supporting Horovod Elastic: https://github.com/kubeflow/mpi-operator/pull/334

Reply:
If we replace kubectl exec with ssh, we can also support the elastic feature, thus I do not think it is a problem for us if I understand it correctly.